paraatm/simulation_method/aviationr.py

```diff
@@ -0,0 +1,210 @@
+
+        self.entityInterface = natsStandalone.getEntityInterface()
+        self.controllerInterface = self.entityInterface.getControllerInterface()
+        self.pilotInterface = self.entityInterface.getPilotInterface()
+
+        self.environmentInterface = natsStandalone.getEnvironmentInterface()
+
+        self.equipmentInterface = natsStandalone.getEquipmentInterface()
+        self.aircraftInterface = self.equipmentInterface.getAircraftInterface()
+
+        if self.simulationInterface is None:
+            natsStandalone.stop()
+            raise RuntimeError("Can't get SimulationInterface")
+
+        self.simulationInterface.clear_trajectory()
+
+    def simulation(self, device, delay=1000, isRNN=True):
+        """
+        Main function for simulation. NATS simulation based commands
+
+        Parameters
+        ----------
+        device : str
+           'CUDA'/ 'CPU'-- Running the model along GPU or CPU
+        delay : int
+            When we start the accident simulation after the required condition meet
+        isRNN : boolean
+            Using RNN or sequential model
+
+        Returns
+        -------
+        dictionary
+            Recorded risk quantification:
+        """
+        # Load data
+        case = pd.read_pickle(self.data + 'case.pickle')
+        case_code = case['case_code']
+        case = case['case']
+        refer = pd.read_pickle(self.data + 'refer.pickle')
+        ntsb2nats = refer['ntsb2nats']
+        datadict = refer['datadict']
+
+        self.simulationInterface.clear_trajectory()
+
+        self.environmentInterface.load_rap(os.environ.get('NATS_HOME') + '/' +"share/tg/rap")
+
+        self.aircraftInterface.load_aircraft(self.fp_file, self.mfl_file)
+
+        self.simulationInterface.setupSimulation(self.sim_time, 30)  # SFO - PHX
+
+        self.simulationInterface.start(1)
+        while True:
+            runtime_sim_status = self.simulationInterface.get_runtime_sim_status()
+            if (runtime_sim_status == self.NATS_SIMULATION_STATUS_PAUSE):
+                break
+
+        # Running the simulation until the phase of aircraft first meet the actual accident phase in NTSB
+        accident = False
+        accident_phase = case.loc[0, 'Phase_of_Flight'] # the initial phase of accident
+        while not accident:
+            self.simulationInterface.resume(delay)
+            while True:
+                runtime_sim_status = self.simulationInterface.get_runtime_sim_status()
+                if (runtime_sim_status == self.NATS_SIMULATION_STATUS_PAUSE):
+                    break
+            aclist = self.aircraftInterface.getAllAircraftId()
+            ac = self.aircraftInterface.select_aircraft(aclist[0])
+            phase = ac.getFlight_phase()
+            if accident_phase in list(ntsb2nats.loc[ntsb2nats['NATS_CODE'] == phase]['NTSB_CODE']): # test if the phase of flight from  NASTS simulation is same as NTSB
+                accident = True
```

Comment:
When I run `aviationr_example.py`, the simulation gets to "Flight propagation completed", and then it hangs with no further progress, but CPU continuing to be used.  Using print statements, it appears that this `while not accident` loop is where it is getting stuck.  Have you come across this before?  How do we ensure that this loop exits?  I am using NATS beta 1.7.

Reply:
Digging into this further, it is actually getting stuck in the inner `while True` loop on line 119.  If I print the value of `runtime_sim_status`, it is 5, which I think corresponds to `RESUME`.  So in my case, it seems to get stuck because the simulation status never comes back as `PAUSE`, so the while loop never exits (unless this simulation just takes a really long time and I am not waiting long enough).

Reply:
I have commit an update for this hanging issue. I think the hanging issue is because the simulation time is too short and my code didn't consider those situation previously.

Reply:
Thanks for the update.  That did make the hanging issue go away, but now the example terminates and says "Accident condition is not met!"  Do you know what might be causing that?

Reply:
Yeah, for this issue, you might need to choose a longer simulation time. In this specific example, you need to set simulation time at least 4500. The logic behind it is that I am trying to map the aviation accident from NTSB to NATS. And for this specific case, the accident is happening during "Cruising" phase which is actually the "condition" that need to be met. 

I first start normal NATS simulation. And when the aircraft enter the "Cruising" phase(It met the accident in NTSB recording), I will start accident simulation. For accident simulation, I still use the NATS simulation but modify the simulation a little bit according to NTSB recording. 

Please let me know if it works and any other problems.

Reply:
I tried the update, but I still get the "Accident condition is not met" message.  I don't really know enough about NATS to be very helpful on this.  Is it possible that the NATS simulation has already passed the phase that you want when you go into the `while not accident` loop?  Is that why the simulation ends before the phase is matched to the accident?  In addition to the simulation time, what about the value that is passed to `simulationInterface.start`?

Reply:
Also, what is the purpose of this:
https://github.com/xz941/para-atm/blob/1d6590a3b40fcac347ed2258da21ec0a7477b044/paraatm/simulation_method/aviationr.py#L109-L112

If I comment that out, then it does get past the loop and find the accident, although then there is a torch/cuda error in the call to `risk_estimator.load_model()`

Reply:
For the value passed in the `simulationInterface.start`, it is the initial simulation time. And I think the code you commented is not necessary. It is pretty strange since I can get 
result when I test the code. Do you want to go over the code through a zoom meeting?

Reply:
For the value passed in the simulationInterface.start, it is the initial simulation time. And I think the code you commented is not necessary which I can removed. It is pretty strange since I can get result when I test the code. Do you want to go over the code through a zoom meeting when you are available?

Reply:
Erin and I spent some time digging into this and were able to recreate the issue also in GNATS using simplified test cases.  The problem has to do with checking for the PAUSE status in NATS.  Without "sleep" commands from Python, NATS will tend to "miss" the pause and run the simulation all the way to the end.  That is why when I run your example, it sometimes does not find the accident (because the simulation is not actually pausing at the specified interval).

After a lot of trial and error, I think that the solution is to add a `time.sleep()` command inside your loop (note that OSI includes sleep commands in their examples).  What worked for me was to add a sleep command as the first line inside the `while not accident` loop, as in:
```python
        while not accident:
            time.sleep(0.1)
```

You will also need to add an `import time` to the top.  As far as I can tell, this resolves the pausing issue (there is still an issue with using torch without a CUDA device, as already noted separately).

Reply:
Got it. Thank you so much for the help! I have make an update for this problem.